# **第三章：資料預處理及預覽**

- **作者：** 歐陽振森 教授
- **單位：** 國立高雄科技大學資管系
- **E-mail：** ouyangcs@nkust.edu.tw
- **版本：** 2025-03-26

# **學習目標**

本章旨在讓學員掌握資料處理與清洗的核心技能，並能夠運用 Pandas、Scikit-learn 等常用工具進行資料預處理。學習目標包含：

- **資料讀取與初探：** 了解如何讀取 CSV 檔案並快速探索資料結構、欄位型態及基本統計量。
- **缺失值與異常值處理：** 學習檢查資料中的缺失與異常值，並運用刪除或填補等方法進行處理。
- **資料型態轉換：** 掌握日期與數值型態轉換，確保後續數據運算正確。
- **資料篩選、排序與分組：** 透過條件過濾、排序與分組統計，深入了解各診斷群體的特徵。
- **資料整合與轉換：** 學會使用合併、重塑等方法整合來自不同資料源的資訊，提升資料結構性。
- **標準化與正規化：** 應用標準化與 Min-Max 正規化技術，消除量綱差異，為後續模型分析做好準備。
- **重複資料檢查與輸出：** 掌握去除重複記錄的方法，並學習將清洗後的資料輸出保存。

透過這些步驟的實作範例，學員將能夠針對資料集進行全面、有效的資料預處理，為後續的視覺化設計或資料建模打下堅實基礎。

### **1. 資料讀取與初探**

**目的：**
- 讀取資料集，快速了解資料基本結構、欄位型態與統計特性，為後續清洗與分析奠定基礎。

**方法：**
- 使用 `pd.read_csv()` 讀取 CSV 檔案
- 利用 `head()` 與 `tail()` 檢視資料的前幾筆與後幾筆
- 使用 `info()` 觀察資料型態與缺失狀況
- 使用 `describe()` 獲取數值欄位的基本統計數據

**範例程式：**

In [ ]:
from google.colab import drive
drive.mount("/content/drive", force_remount=True)
path = '/content/drive/MyDrive/Colab Notebooks/CH3_dataset/dataset'

Mounted at /content/drive


In [ ]:
import pandas as pd

# 讀取醫療資料集
from pathlib import Path

file_name = "medical_records.csv"
df = pd.read_csv(path +"/"+ file_name)

# 初步檢視資料前五筆與後五筆
print("資料前五筆：")
print(df.head())
print("\n資料後五筆：")
print(df.tail())

# 檢查資料型態與缺失值情形
print("\n資料摘要：")
print(df.info())

# 查看數值欄位的統計量
print("\n數值欄位統計量：")
print(df.describe())

資料前五筆：
  PatientID      Name   Age     Diagnosis  BloodPressure AdmissionDate
0      P001  Patient1  55.0  Hypertension          130.0    2023-01-01
1      P002  Patient2  42.0      Diabetes          125.0    2023-01-02
2      P003  Patient3  68.0        Asthma          140.0    2023-01-03
3      P004  Patient4  30.0           Flu          110.0    2023-01-04
4      P005  Patient5   NaN  Hypertension          135.0    2023-01-05

資料後五筆：
   PatientID       Name   Age     Diagnosis  BloodPressure AdmissionDate
45      P046  Patient46  67.0        Asthma          134.0    2023-02-15
46      P047  Patient47  54.0           Flu          129.0    2023-02-16
47      P048  Patient48  72.0  Hypertension          131.0    2023-02-17
48      P049  Patient49  61.0      Diabetes          138.0    2023-02-18
49      P050  Patient50  59.0           NaN          133.0    2023-02-19

資料摘要：
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 6 columns):
 #   Column 

### Try it!：1. 資料讀取與初探
你是醫院的資料分析師，剛拿到一份病歷資料，需要先確認資料的基本狀況。

**程式要求**
1. 讀取 `dataset/medical_records.csv`。
2. 印出前 5 筆資料、資料筆數與欄位數、每個欄位的資料型態。
3. 建立字典 `basic_report`，包含 `rows`、`columns`、`missing_total`（全部缺失值總數），並印出。

### **2. 缺失值與異常值處理**

**目的：**
- 檢查資料中的缺失值與可能存在的異常值，並採取適當的處理策略以提升資料品質。

**方法：**
- 使用 `isnull()` 檢查每個欄位的缺失值
- 透過 `dropna()` 刪除關鍵欄位的缺失值（例如診斷資訊）
- 利用 `fillna()` 以均值或中位數填補數值欄位缺失值

**範例程式：**

In [ ]:
# 檢查缺失值情形
print("\n各欄位缺失值數量：")
print(df.isnull().sum())

# 對數值欄位以均值/中位數填補缺失值
df['Age'] = df['Age'].fillna(df['Age'].mean())
df['BloodPressure'] = df['BloodPressure'].fillna(df['BloodPressure'].median())

# 對診斷欄位若缺失則刪除該筆記錄
df = df.dropna(subset=['Diagnosis'])

# 此外，利用描述統計或視覺化（此處僅示例）檢查異常值
print("\n更新後各欄位缺失值數量：")
print(df.isnull().sum())


各欄位缺失值數量：
PatientID        0
Name             0
Age              3
Diagnosis        5
BloodPressure    3
AdmissionDate    0
dtype: int64

更新後各欄位缺失值數量：
PatientID        0
Name             0
Age              0
Diagnosis        0
BloodPressure    0
AdmissionDate    0
dtype: int64


### Try it!：2. 缺失值與異常值處理
病歷資料中有部分欄位缺漏，血壓也可能出現量測異常，需要先清理才能分析。

**程式要求**
1. 讀取 `dataset/medical_records.csv`，複製一份為 `df_clean`。
2. 以中位數填補 `Age` 與 `BloodPressure` 的缺失值。
3. 移除 `BloodPressure` 大於 200 或小於 50 的資料列，印出處理前後的資料筆數。

### **3. 資料類型轉換**

**目的：**
- 將資料轉換為適合分析的數據型態，如日期轉換與數值型態轉換，確保後續運算正確。

**方法：**
- 使用 `pd.to_datetime()` 將日期字串轉換成 datetime 物件
- 使用 `astype()` 將數值欄位（如年齡）轉換成整數或其他適當型態

**範例程式：**

In [ ]:
# 轉換 AdmissionDate 欄位為日期型態
df['AdmissionDate'] = pd.to_datetime(df['AdmissionDate'])

# 將 Age 欄位轉換為整數型態
df['Age'] = df['Age'].astype(int)

# 檢查轉換結果
print("\n轉換後的 AdmissionDate 與 Age 資料型態：")
print(df[['AdmissionDate', 'Age']].dtypes)


轉換後的 AdmissionDate 與 Age 資料型態：
AdmissionDate    datetime64[ns]
Age                       int64
dtype: object


### Try it!：3. 資料類型轉換
入院日期目前是文字格式，需要轉成日期型態才能做時間分析。

**程式要求**
1. 讀取 `dataset/medical_records.csv`，將 `AdmissionDate` 轉為 datetime 型態。
2. 新增 `AdmissionYear` 與 `AdmissionMonth` 兩個欄位。
3. 將 `Age` 填補缺失值後轉為整數型態，印出 `df.dtypes` 檢查結果。

### **4. 資料篩選與過濾**

**目的：**
- 根據特定條件篩選出目標資料（例如特定年齡或診斷群體），以便進一步分析。

**方法：**
- 使用布林索引進行條件過濾
- 使用 `query()` 方法根據條件篩選資料

**範例程式：**

In [ ]:
# 篩選出年齡大於 50 且診斷為「Hypertension」（高血壓）的病患
df_filtered = df[(df['Age'] > 50) & (df['Diagnosis'] == 'Hypertension')]
print("\n布林索引篩選結果：")
print(df_filtered.head())

# 或者利用 query() 方法達成同樣的篩選條件
df_filtered_query = df.query("Age > 50 and Diagnosis == 'Hypertension'")
print("\nquery() 篩選結果：")
print(df_filtered_query.head())


布林索引篩選結果：
   PatientID       Name  Age     Diagnosis  BloodPressure AdmissionDate
0       P001   Patient1   55  Hypertension          130.0    2023-01-01
4       P005   Patient5   59  Hypertension          135.0    2023-01-05
8       P009   Patient9   65  Hypertension          145.0    2023-01-09
13      P014  Patient14   60  Hypertension          136.0    2023-01-14
17      P018  Patient18   75  Hypertension          141.0    2023-01-18

query() 篩選結果：
   PatientID       Name  Age     Diagnosis  BloodPressure AdmissionDate
0       P001   Patient1   55  Hypertension          130.0    2023-01-01
4       P005   Patient5   59  Hypertension          135.0    2023-01-05
8       P009   Patient9   65  Hypertension          145.0    2023-01-09
13      P014  Patient14   60  Hypertension          136.0    2023-01-14
17      P018  Patient18   75  Hypertension          141.0    2023-01-18


### Try it!：4. 資料篩選與過濾
醫師想找出需要優先追蹤的糖尿病高風險病患。

**程式要求**
1. 讀取 `dataset/medical_records.csv`，篩選出 `Diagnosis` 為 `Diabetes` 的病患。
2. 從中再篩選 `Age >= 50` 且 `BloodPressure >= 130` 的資料列。
3. 印出篩選結果。

### **5. 資料排序與分組統計**

**目的：**
- 對資料依據特定欄位進行排序，並分組計算各組別的統計數據，協助了解不同診斷群體的特徵。

**方法：**
- 使用 `sort_values()` 依據 AdmissionDate 進行排序
- 使用 `groupby()` 結合聚合函數（如 mean、sum、count）進行分組統計

**範例程式：**

In [ ]:
# 依據 AdmissionDate 進行排序（由舊至新）
df_sorted = df.sort_values(by='AdmissionDate', ascending=True)
print("\n依 AdmissionDate 排序後的資料：")
print(df_sorted.head())

# 以 Diagnosis 分組，計算各組別平均年齡與平均血壓
grouped = df.groupby('Diagnosis').agg({
    'Age': 'mean',
    'BloodPressure': 'mean'
}).reset_index()
print("\n以 Diagnosis 分組的統計結果：")
print(grouped)


依 AdmissionDate 排序後的資料：
  PatientID      Name  Age     Diagnosis  BloodPressure AdmissionDate
0      P001  Patient1   55  Hypertension          130.0    2023-01-01
1      P002  Patient2   42      Diabetes          125.0    2023-01-02
2      P003  Patient3   68        Asthma          140.0    2023-01-03
3      P004  Patient4   30           Flu          110.0    2023-01-04
4      P005  Patient5   59  Hypertension          135.0    2023-01-05

以 Diagnosis 分組的統計結果：
      Diagnosis        Age  BloodPressure
0        Asthma  58.545455     129.636364
1      Diabetes  58.833333     136.166667
2           Flu  57.900000     130.900000
3  Hypertension  65.083333     138.166667


### Try it!：5. 資料排序與分組統計
院方想了解不同診斷類別病患的年齡與血壓概況。

**程式要求**
1. 讀取 `dataset/medical_records.csv`，依 `Age` 由大到小排序並印出前幾筆。
2. 使用 `groupby('Diagnosis')` 計算各診斷類別的平均年齡與平均血壓。
3. 印出分組統計結果。

### **6. 資料整合與轉換**

**目的：**
- 將來自不同來源的資料合併，或對同一資料集進行重塑，提升資料的結構性與可讀性。

**方法：**
- 使用 `merge()` 合併多個 DataFrame
- 使用 `pivot_table()` 重塑資料表結構
- 使用 `concat()` 合併資料列或欄

**範例程式：**

In [ ]:
# 假設利用同一資料集拆分出基本資料與檢查結果，再進行合併

# 提取基本資料子集（PatientID、Name、Age、Diagnosis）
df_basic = df[['PatientID', 'Name', 'Age', 'Diagnosis']]

# 提取檢查結果子集（PatientID、BloodPressure、AdmissionDate）
df_vitals = df[['PatientID', 'BloodPressure', 'AdmissionDate']]

# 以 PatientID 為關鍵欄位進行合併
df_merged = pd.merge(df_basic, df_vitals, on='PatientID', how='inner')
print("\n合併後的資料：")
print(df_merged.head())

# 利用 pivot_table 重塑資料，計算各診斷群組的平均血壓
pivot = pd.pivot_table(df_merged, values='BloodPressure', index='Diagnosis', aggfunc='mean')
print("\nPivot Table 統計各診斷群組平均血壓：")
print(pivot)


合併後的資料：
  PatientID      Name  Age     Diagnosis  BloodPressure AdmissionDate
0      P001  Patient1   55  Hypertension          130.0    2023-01-01
1      P002  Patient2   42      Diabetes          125.0    2023-01-02
2      P003  Patient3   68        Asthma          140.0    2023-01-03
3      P004  Patient4   30           Flu          110.0    2023-01-04
4      P005  Patient5   59  Hypertension          135.0    2023-01-05

Pivot Table 統計各診斷群組平均血壓：
              BloodPressure
Diagnosis                  
Asthma           129.636364
Diabetes         136.166667
Flu              130.900000
Hypertension     138.166667


### Try it!：6. 資料整合與轉換
換一份電商訂單資料，練習用樞紐分析表整理每月銷售概況。

**程式要求**
1. 讀取 `dataset/ecommerce_orders.csv`，將 `OrderDate` 轉為日期後建立 `OrderMonth` 欄位。
2. 使用 `pivot_table` 計算各 `ProductCategory` 每月 `OrderAmount` 總和。
3. 印出樞紐分析結果。

### **7. 資料標準化與正規化**

**目的：**
- 進行數值資料的標準化與正規化處理，以消除量綱差異，方便後續模型建立與分析。

**方法：**
- 使用 `sklearn.preprocessing.StandardScaler` 進行標準化
- 手動計算並實現 Min-Max 正規化

**範例程式：**

In [ ]:
from sklearn.preprocessing import StandardScaler

# 標準化 BloodPressure 欄位
scaler = StandardScaler()
df['BloodPressure_scaled'] = scaler.fit_transform(df[['BloodPressure']])
print("\n標準化後的 BloodPressure 前五筆：")
print(df[['BloodPressure', 'BloodPressure_scaled']].head())

# 手動進行 Age 欄位的 Min-Max 正規化
df['Age_norm'] = (df['Age'] - df['Age'].min()) / (df['Age'].max() - df['Age'].min())
print("\n正規化後的 Age 前五筆：")
print(df[['Age', 'Age_norm']].head())


標準化後的 BloodPressure 前五筆：
   BloodPressure  BloodPressure_scaled
0          130.0             -0.426418
1          125.0             -0.968475
2          140.0              0.657696
3          110.0             -2.594646
4          135.0              0.115639

正規化後的 Age 前五筆：
   Age  Age_norm
0   55  0.454545
1   42  0.218182
2   68  0.690909
3   30  0.000000
4   59  0.527273


### Try it!：7. 資料標準化與正規化
建模前常需要把不同量級的數值欄位轉換到相近的尺度。

**程式要求**
1. 讀取 `dataset/medical_records_cleaned.csv`。
2. 使用 `StandardScaler` 產生 `BloodPressure_scaled_new` 欄位。
3. 使用 min-max 公式產生 `Age_norm_new` 欄位，印出兩個新欄位的前 10 筆。

### **8. 檢查與處理重複資料**

**目的：**
- 檢查資料中是否存在重複記錄，並透過去重確保資料唯一性，避免分析偏差。

**方法：**
- 使用 `duplicated()` 檢查重複資料
- 使用 `drop_duplicates()` 刪除重複記錄

**範例程式：**

In [ ]:
# 檢查重複記錄數量
num_duplicates = df.duplicated().sum()
print(f"\n重複記錄數量： {num_duplicates}")

# 刪除重複記錄
df = df.drop_duplicates()
print("\n刪除重複記錄後，資料筆數：", len(df))


重複記錄數量： 0

刪除重複記錄後，資料筆數： 45


### Try it!：8. 檢查與處理重複資料
訂單資料可能因重複匯入而出現相同的訂單編號。

**程式要求**
1. 讀取 `dataset/ecommerce_orders.csv`。
2. 使用 `duplicated()` 檢查 `OrderID` 是否重複，印出重複筆數。
3. 以 `drop_duplicates()` 移除重複訂單，顯示處理後的資料。

### **9. 數據輸出與存檔**

**目的：**
- 將清洗與處理後的資料儲存為新的檔案，供後續分析或視覺化使用。

**方法：**
- 使用 `to_csv()` 將 DataFrame 輸出為 CSV 檔案
- （選擇性）使用 `to_excel()` 輸出為 Excel 檔案

**範例程式：**

In [ ]:
# 輸出清洗後的資料至新的 CSV 檔案
df.to_csv("medical_records_cleaned.csv", index=False)
print("\n資料已成功輸出為 'medical_records_cleaned.csv'")


資料已成功輸出為 'medical_records_cleaned.csv'


### Try it!：9. 數據輸出與存檔
把前面學到的步驟串成一次完整的清理流程。

**程式要求**
1. 讀取 `dataset/medical_records.csv`，以中位數填補 `Age` 與 `BloodPressure` 的缺失值。
2. 將 `AdmissionDate` 轉為 datetime，並移除重複資料列。
3. 顯示清理後的資料與最終筆數。

### **綜合範例程式**

下列程式碼整合前述各章節步驟，採用同一份醫療資料集 **medical_records.csv**，展示從資料讀取、清洗、轉換、篩選、排序、合併、標準化、去重到最後輸出的完整流程：

In [ ]:
import pandas as pd
from sklearn.preprocessing import StandardScaler

# === 1. 數據讀取與初探 ===
# Kaggle API
import kagglehub

# Download latest version
path = kagglehub.dataset_download("f114118114/ch2-dataset")

print("Path to dataset files:", path)

from pathlib import Path

file_name = "medical_records.csv"
df = pd.read_csv(path +"/"+ file_name)
print("【初探資料】")
print(df.head())
print(df.info())
print(df.describe())

# === 2. 缺失值與異常值處理 ===
print("\n【缺失值檢查】")
print(df.isnull().sum())

df['Age'] = df['Age'].fillna(df['Age'].mean())
df['BloodPressure'] = df['BloodPressure'].fillna(df['BloodPressure'].median())
df = df.dropna(subset=['Diagnosis'])
print("缺失值處理後：")
print(df.isnull().sum())

# === 3. 資料型態轉換 ===
df['AdmissionDate'] = pd.to_datetime(df['AdmissionDate'])
df['Age'] = df['Age'].astype(int)
print("\n【資料型態轉換】")
print(df[['AdmissionDate', 'Age']].dtypes)

# === 4. 資料篩選與過濾 ===
df_filtered = df[(df['Age'] > 50) & (df['Diagnosis'] == 'Hypertension')]
print("\n【布林索引篩選】")
print(df_filtered.head())

df_filtered_query = df.query("Age > 50 and Diagnosis == 'Hypertension'")
print("【query() 篩選】")
print(df_filtered_query.head())

# === 5. 資料排序與分組統計 ===
df_sorted = df.sort_values(by='AdmissionDate', ascending=True)
print("\n【依 AdmissionDate 排序】")
print(df_sorted.head())

grouped = df.groupby('Diagnosis').agg({'Age': 'mean', 'BloodPressure': 'mean'}).reset_index()
print("\n【分組統計】")
print(grouped)

# === 6. 資料整合與轉換 ===
df_basic = df[['PatientID', 'Name', 'Age', 'Diagnosis']]
df_vitals = df[['PatientID', 'BloodPressure', 'AdmissionDate']]
df_merged = pd.merge(df_basic, df_vitals, on='PatientID', how='inner')
print("\n【合併資料】")
print(df_merged.head())

pivot = pd.pivot_table(df_merged, values='BloodPressure', index='Diagnosis', aggfunc='mean')
print("\n【Pivot Table】")
print(pivot)

# === 7. 資料標準化與正規化 ===
scaler = StandardScaler()
df['BloodPressure_scaled'] = scaler.fit_transform(df[['BloodPressure']])
print("\n【標準化】")
print(df[['BloodPressure', 'BloodPressure_scaled']].head())

df['Age_norm'] = (df['Age'] - df['Age'].min()) / (df['Age'].max() - df['Age'].min())
print("【正規化】")
print(df[['Age', 'Age_norm']].head())

# === 8. 檢查與處理重複資料 ===
num_duplicates = df.duplicated().sum()
print(f"\n重複記錄數量： {num_duplicates}")
df = df.drop_duplicates()
print("去重後資料筆數：", len(df))

# === 9. 數據輸出與存檔 ===
df.to_csv("medical_records_cleaned.csv", index=False)
print("\n資料已成功輸出為 'medical_records_cleaned.csv'")

100%|██████████| 3.76k/3.76k [00:00<00:00, 7.66MB/s]

Extracting files...
Path to dataset files: /root/.cache/kagglehub/datasets/f114118114/ch2-dataset/versions/1
【初探資料】
  PatientID      Name   Age     Diagnosis  BloodPressure AdmissionDate
0      P001  Patient1  55.0  Hypertension          130.0    2023-01-01
1      P002  Patient2  42.0      Diabetes          125.0    2023-01-02
2      P003  Patient3  68.0        Asthma          140.0    2023-01-03
3      P004  Patient4  30.0           Flu          110.0    2023-01-04
4      P005  Patient5   NaN  Hypertension          135.0    2023-01-05
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 6 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PatientID      50 non-null     object 
 1   Name           50 non-null     object 
 2   Age            47 non-null     float64
 3   Diagnosis      45 non-null     object 
 4   BloodPressure  47 non-null     float64
 5   AdmissionDate  50 non-null     object 
dtyp